mlflow/sklearn/__init__.py

```diff
@@ -994,3 +986,102 @@ def patched_fit(original, self, *args, **kwargs):
                 safe_patch(
                     FLAVOR_NAME, class_def, func_name, patched_fit, manage_run=True,
                 )
+
+
+def log_eval_metrics(*, model=None, model_uri=None, X, y_true, prefix, sample_weight=None):
+    """
+    Computes and logs evaluation metrics (and artifacts) for the given model and labeled dataset.
+    The metrics/artifacts mirror what is auto-logged when training a model
+    (see mlflow.sklearn.autolog).
+
+    ** Example **
+
+    .. code-block:: python
+
+        from sklearn.linear_model import LinearRegression
+        import mlflow
+
+        # enable autologging
+        mlflow.sklearn.autolog()
+
+        # prepare training data
+        X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
+        y = np.dot(X, np.array([1, 2])) + 3
+
+        # prepare evaluation data
+        X_eval = np.array([[3, 3], [3, 4]])
+        y_eval = np.dot(X_eval, np.array([1,2])) + 3
+
+        # train a model
+        model = LinearRegression()
+        with mlflow.start_run() as run:
+            model.fit(X, y)
+            mlflow.log_eval_metrics(model=model, X=X_eval, y_true=y_eval, prefix="val_")
+
+
+    Each metric's and artifact's name is prefixed with `prefix`, e.g., in the previous example the
+    metrics and artifacts are named 'eval_XXXXX'. Note that training-time metrics are auto-logged
+    as 'training_XXXXX'.
+
+    The run under which to log the metrics/artifacts is chosen as follows:
+    - Under the active run if one exists
+    - Or, under the same run as the model_uri if the latter is of the form 'runs:/...'
```

Comment:
not sure this is what users want. maybe easier to require an active run.

Reply:
Always requiring an active run seems to go against the mlflow APIs (e.g., mlflow.log_metric() will initialize a new run if one is not present). So this leaves the question of what to do when there is no active run and the user specifies a runs:/ URL. We can avoid this for now if we decide to not support model_uri. However, if we support it then I can see a case for re-using the same run to log more metrics/artifacts for different datasets, so that everything is available in the same run that generated the model. The flip side is that it makes the handling of model_uri is inconsistent, and this might indeed be surprising to the users.

Reply:
Done.